File: plugins/agentic-sdlc/src/agentic_sdlc/steps/conditional_step.py
"""Conditional step executor."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from agentic_sdlc.progress import WorkflowStatus, update_step_completed, update_step_failed
from agentic_sdlc.steps.base import StepContext, StepExecutor, StepResult

if TYPE_CHECKING:
    from agentic_sdlc.console import ConsoleOutput
    from agentic_sdlc.logging.logger import WorkflowLogger
    from agentic_sdlc.parser import StepDefinition
    from agentic_sdlc.progress import WorkflowProgress
    from agentic_sdlc.steps.parallel_step import BranchStepExecutor


class ConditionalStepExecutor(StepExecutor):
    """Executes a conditional step based on condition evaluation.

    Evaluates the condition expression and executes either the 'then' or 'else' branch.
    """

    def __init__(self, branch_executor: BranchStepExecutor):
        """Initialize with branch executor for nested steps.

        Args:
            branch_executor: Executor for steps within conditional branches
        """
        self.branch_executor = branch_executor

    def execute(
        self,
        step: StepDefinition,
        progress: WorkflowProgress,
        context: StepContext,
        logger: WorkflowLogger,
        console: ConsoleOutput,
    ) -> StepResult:
        """Execute a conditional step based on condition evaluation."""
        condition = step.condition or "false"
        logger.info(step.name, f"Evaluating condition: {condition}")

        template_context = context.build_template_context()
        if context.renderer.has_variables(condition):
            condition = context.renderer.render_string(condition, template_context)

        condition_result = self._evaluate_condition(condition, template_context)
        logger.info(step.name, f"Condition evaluated to: {condition_result}")
        console.info(f"Conditional '{step.name}': {condition} = {condition_result}")

        if condition_result:
            steps_to_run = step.then_steps
            branch_name = "then"
        else:
            steps_to_run = step.else_steps
            branch_name = "else"

        if not steps_to_run:
            output_summary = f"Condition {condition_result}, no '{branch_name}' branch to execute"
            update_step_completed(progress, step.name, output_summary)
            console.step_complete(step.name, output_summary)
            return StepResult(success=True, output_summary=output_summary)

        logger.info(step.name, f"Executing '{branch_name}' branch with {len(steps_to_run)} steps")

        for sub_step in steps_to_run:
            try:
                result = self.branch_executor.execute(sub_step, progress, context, logger, console)

                if not result.success:
                    logger.warning(step.name, f"Conditional '{branch_name}' branch stopped due to failure")
                    error_msg = f"Step '{sub_step.name}' failed: {result.error}"
                    update_step_failed(progress, step.name, error_msg)
                    console.step_failed(step.name, f"Step '{sub_step.name}' failed")
                    progress.status = WorkflowStatus.FAILED.value
                    return StepResult(success=False, error=error_msg)

                if progress.status == WorkflowStatus.FAILED.value:
                    logger.warning(step.name, f"Conditional '{branch_name}' branch stopped due to failure")
                    break

            except Exception as e:
                logger.error(sub_step.name, f"Step failed in conditional branch: {e}")
                error_msg = f"Step '{sub_step.name}' failed: {e}"
                update_step_failed(progress, step.name, error_msg)
                console.step_failed(step.name, f"Step '{sub_step.name}' failed")
                progress.status = WorkflowStatus.FAILED.value
                return StepResult(success=False, error=error_msg)

        if progress.status != WorkflowStatus.FAILED.value:
            output_summary = f"Executed '{branch_name}' branch ({len(steps_to_run)} steps)"
            update_step_completed(progress, step.name, output_summary)
            console.step_complete(step.name, output_summary)
            logger.info(step.name, output_summary)
            return StepResult(success=True, output_summary=output_summary)

        return StepResult(success=False, error="Conditional execution failed")

    def _evaluate_condition(self, condition: str, context: dict[str, Any]) -> bool:
        """Evaluate a condition expression.

        Supports simple comparisons and variable access:
        - variables.name == 'value'
        - variables.name != 'value'
        - variables.flag (truthy check)
        """
        condition = condition.strip()

        if condition.lower() in ("true", "1", "yes"):
            return True
        if condition.lower() in ("false", "0", "no", "none", ""):
            return False

        if "!=" in condition:
            left, right = condition.split("!=", 1)
            left_val = self._resolve_value(left.strip(), context)
            right_val = self._resolve_value(right.strip(), context)
            return left_val != right_val

        if "==" in condition:
            left, right = condition.split("==", 1)
            left_val = self._resolve_value(left.strip(), context)
            right_val = self._resolve_value(right.strip(), context)
            return left_val == right_val

        value = self._resolve_value(condition, context)
        return bool(value) and value not in ("none", "None", "null", "")

    def _resolve_value(self, expr: str, context: dict[str, Any]) -> Any:
        """Resolve a value expression from context or as a literal."""
        expr = expr.strip()

        if (expr.startswith("'") and expr.endswith("'")) or (expr.startswith('"') and expr.endswith('"')):
            return expr[1:-1]

        if expr.startswith("variables."):
            var_name = expr[10:]
            return context.get("variables", {}).get(var_name)

        if expr.startswith("outputs."):
            output_name = expr[8:]
            return context.get("outputs", {}).get(output_name)

        return context.get(expr, expr)

File: src/steps/conditional-step.ts
/** Conditional step executor. */

import type { ConsoleOutput } from "../console.js";
import type { WorkflowLogger } from "../logging/logger.js";
import { WORKFLOW_STATUS, updateStepCompleted, updateStepFailed } from "../progress.js";
import type { StepDefinition, WorkflowProgress } from "../types.js";
import {
	type BranchStepExecutor,
	type StepContext,
	StepExecutor,
	type StepResult,
	buildTemplateContext,
} from "./base.js";

export class ConditionalStepExecutor extends StepExecutor {
	constructor(private branchExecutor: BranchStepExecutor) {
		super();
	}

	async execute(
		step: StepDefinition,
		progress: WorkflowProgress,
		context: StepContext,
		logger: WorkflowLogger,
		console: ConsoleOutput,
	): Promise<StepResult> {
		let condition = step.condition ?? "false";
		logger.info(step.name, `Evaluating condition: ${condition}`);

		const templateContext = buildTemplateContext(context);
		if (context.renderer.hasVariables(condition)) {
			condition = context.renderer.renderString(condition, templateContext);
		}

		const conditionResult = this.evaluateCondition(condition, templateContext);
		logger.info(step.name, `Condition evaluated to: ${conditionResult}`);
		console.info(`Conditional '${step.name}': ${condition} = ${conditionResult}`);

		const stepsToRun = conditionResult ? step.thenSteps : step.elseSteps;
		const branchName = conditionResult ? "then" : "else";

		if (!stepsToRun || stepsToRun.length === 0) {
			const outputSummary = `Condition ${conditionResult}, no '${branchName}' branch to execute`;
			updateStepCompleted(progress, step.name, outputSummary);
			console.stepComplete(step.name, outputSummary);
			return { success: true, outputSummary };
		}

		logger.info(step.name, `Executing '${branchName}' branch with ${stepsToRun.length} steps`);

		for (const subStep of stepsToRun) {
			try {
				const result = await this.branchExecutor(subStep, progress, context, logger, console);

				if (!result.success) {
					logger.warning(step.name, `Conditional '${branchName}' branch stopped due to failure`);
					const errorMsg = `Step '${subStep.name}' failed: ${result.error}`;
					updateStepFailed(progress, step.name, errorMsg);
					console.stepFailed(step.name, `Step '${subStep.name}' failed`);
					progress.status = WORKFLOW_STATUS.FAILED;
					return { success: false, error: errorMsg };
				}

				if (progress.status === WORKFLOW_STATUS.FAILED) {
					logger.warning(step.name, `Conditional '${branchName}' branch stopped due to failure`);
					break;
				}
			} catch (e: unknown) {
				const errStr = e instanceof Error ? e.message : String(e);
				logger.error(subStep.name, `Step failed in conditional branch: ${errStr}`);
				const errorMsg = `Step '${subStep.name}' failed: ${errStr}`;
				updateStepFailed(progress, step.name, errorMsg);
				console.stepFailed(step.name, `Step '${subStep.name}' failed`);
				progress.status = WORKFLOW_STATUS.FAILED;
				return { success: false, error: errorMsg };
			}
		}

		if (progress.status !== WORKFLOW_STATUS.FAILED) {
			const outputSummary = `Executed '${branchName}' branch (${stepsToRun.length} steps)`;
			updateStepCompleted(progress, step.name, outputSummary);
			console.stepComplete(step.name, outputSummary);
			logger.info(step.name, outputSummary);
			return { success: true, outputSummary };
		}

		return { success: false, error: "Conditional execution failed" };
	}

	evaluateCondition(condition: string, context: Record<string, unknown>): boolean {
		const trimmed = condition.trim();

		if (["true", "1", "yes"].includes(trimmed.toLowerCase())) {
			return true;
		}
		if (["false", "0", "no", "none", ""].includes(trimmed.toLowerCase())) {
			return false;
		}

		if (trimmed.includes("!=")) {
			const [left, right] = trimmed.split("!=", 2);
			const leftVal = this.resolveValue(left.trim(), context);
			const rightVal = this.resolveValue(right.trim(), context);
			return leftVal !== rightVal;
		}

		if (trimmed.includes("==")) {
			const [left, right] = trimmed.split("==", 2);
			const leftVal = this.resolveValue(left.trim(), context);
			const rightVal = this.resolveValue(right.trim(), context);
			return leftVal === rightVal;
		}

		const value = this.resolveValue(trimmed, context);
		return (
			Boolean(value) && value !== "none" && value !== "None" && value !== "null" && value !== ""
		);
	}

	resolveValue(expr: string, context: Record<string, unknown>): unknown {
		const trimmed = expr.trim();

		if (
			(trimmed.startsWith("'") && trimmed.endsWith("'")) ||
			(trimmed.startsWith('"') && trimmed.endsWith('"'))
		) {
			return trimmed.slice(1, -1);
		}

		if (trimmed.startsWith("variables.")) {
			const varName = trimmed.slice(10);
			const vars = context.variables as Record<string, unknown> | undefined;
			return vars?.[varName];
		}

		if (trimmed.startsWith("outputs.")) {
			const outputName = trimmed.slice(8);
			const outputs = context.outputs as Record<string, unknown> | undefined;
			return outputs?.[outputName];
		}

		return (context as Record<string, unknown>)[trimmed] ?? trimmed;
	}
}
